OpenRA.Game/Server/VoteKickTracker.cs

```diff
@@ -0,0 +1,178 @@
+				return;
+
+			if (!server.Conns.Contains(kickee.Conn))
+			{
+				server.SendLocalizedMessage(VoteKickEnded, Translation.Arguments("kickee", kickee.Client.Name));
+				EndKickVote();
+				return;
+			}
+
+			if (voteKickTimer.ElapsedMilliseconds > server.Settings.VoteKickTimer)
+			{
+				server.SendLocalizedMessage(VoteKickEnded, Translation.Arguments("kickee", kickee.Client.Name));
+				EndKickVoteAndBlockKicker();
+			}
+		}
+
+		public bool VoteKick(Connection conn, Session.Client kicker, Connection kickeeConn, Session.Client kickee, int kickeeID, bool vote)
+		{
+			if (server.State != ServerState.GameStarted
+				|| (kickee.IsAdmin && server.Type != ServerType.Dedicated)
+				|| (!vote && this.kickee.Client == null) // Disallow starting a vote with a downvote
+				|| (this.kickee.Client != null && this.kickee.Client != kickee) // Disallow starting new votes when one is already ongoing.
+				|| !ClientHasPower(kicker))
+			{
+				server.SendLocalizedMessageTo(conn, UnableToStartAVote);
+				return false;
+			}
+
+			// Prevent vote kick spam abuse.
+			if (voteKickTimer == null && failedVoteKickers.TryGetValue(kicker, out var time))
+			{
+				if (time + server.Settings.VoteKickerCooldown > kickeeConn.ConnectionTimer.ElapsedMilliseconds)
+				{
+					server.SendLocalizedMessageTo(conn, UnableToStartAVote);
+					return false;
+				}
+				else
+					failedVoteKickers.Remove(kicker);
+			}
+
+			var eligiblePlayers = server.Conns.Count(conn => ClientHasPower(server.GetClient(conn)));
+
+			// If only 2 players are playing, they should not be able to kick eachother.
+			if (!kickee.IsObserver && eligiblePlayers < 3)
+			{
+				server.SendLocalizedMessageTo(conn, InsufficientVotes, Translation.Arguments("kickee", kickee.Name));
+				return false;
+			}
+
+			if (voteKickTimer == null)
+			{
+				Log.Write("server", $"Vote kick started on {kickeeID}.");
+				voteKickTimer = Stopwatch.StartNew();
+				server.SendLocalizedMessage(VoteKickStarted, Translation.Arguments("kicker", kicker.Name, "kickee", kickee.Name));
+				server.DispatchServerOrdersToClients(new Order("StartKickVote", null, false) { ExtraData = (uint)kickeeID }.Serialize());
+				this.kickee = (kickee, kickeeConn);
+				voteKickerStarter = (kicker, conn);
+			}
+
+			if (!voteTracker.ContainsKey(conn.PlayerIndex))
+				voteTracker[conn.PlayerIndex] = vote;
+			else
+			{
+				server.SendLocalizedMessageTo(conn, AlreadyVoted, null);
+				return false;
+			}
+
+			var votesFor = 0;
+			var votesAgainst = 0;
+			foreach (var c in voteTracker)
+			{
+				if (c.Value)
+					votesFor++;
+				else
+					votesAgainst++;
+			}
+
+			var percentage = votesFor * 100 / eligiblePlayers;
+			server.SendLocalizedMessage(VoteKickProgress, Translation.Arguments("kickee", kickee.Name, "percentage", percentage));
+
+			// If only a single player is playing, allow him to kick anyone.
+			if (percentage > 50 || (kickee.IsObserver && kickee.IsAdmin && eligiblePlayers < 3))
```

Comment:
Why does this need to check `kickee.IsAdmin`?

Reply:
changed the check